server/cluster_worker.go

```
@@ -173,23 +173,55 @@ func (c *raftCluster) updateJobStatus(job *pd_jobpd.Job, status pd_jobpd.JobStat
 	return nil
 }
 
+type checkOKFunc func(*raft_cmdpb.RaftCmdRequest) (*raft_cmdpb.AdminResponse, error)
+
 func (c *raftCluster) handleJob(job *pd_jobpd.Job) error {
 	log.Debugf("begin to handle job %v", job)
 
-	firstRunning := false
+	var (
+		request = job.Request
+		// must administrator request, check later.
+		adminRequest = request.AdminRequest
+
+		checkOK checkOKFunc
+
+		resp *raft_cmdpb.AdminResponse
+		err  error
+	)
+
+	switch adminRequest.GetCmdType() {
+	case raft_cmdpb.AdminCmdType_Split:
+		checkOK = c.checkSplitOK
+	case raft_cmdpb.AdminCmdType_ChangePeer:
+		checkOK = c.checkChangePeerOK
+	default:
+		log.Errorf("unsupported request %v, ignore", request)
+		return nil
+	}
 
-	// TODO: if the job status is running, check this job whether
-	// finished or not in raft server.
 	if job.GetStatus() == pd_jobpd.JobStatus_Pending {
-		firstRunning = true
-		if err := c.updateJobStatus(job, pd_jobpd.JobStatus_Running); err != nil {
+		if err = c.updateJobStatus(job, pd_jobpd.JobStatus_Running); err != nil {
+			return errors.Trace(err)
+		}
+		// If the job is first running, no need to check whether the job
+		// is finished OK.
+		checkOK = nil
+	} else {
+		// Here means the job is not first running, we can first check whether
+		// the job is finished OK.
+		// The got response != nil means we have already executed the job and
+		// the region version/conf version is changed, so we don't need to process
+		// the job again.
+		if resp, err = checkOK(request); err != nil {
 			return errors.Trace(err)
 		}
 	}
 
-	resp, err := c.processJob(job, firstRunning)
-	if err != nil || resp == nil {
-		return errors.Trace(err)
+	if resp == nil {
+		resp, err = c.processJob(job, checkOK)
+		if err != nil || resp == nil {
+			return errors.Trace(err)
```

If both err and resp equals nil, we should return an error instead of nil.